### PR TITLE
feat: Add support for deletion delay when using PodGC

### PR DESCRIPTION
**Pull request by stefansedich**
The PR https://github.com/argoproj/argo-workflows/pull/5033 was meant to fix the issue of delaying pod deletion for some period so things like logging scrapers can do their thing before the pod goes away.

Based on my comments https://github.com/argoproj/argo-workflows/pull/5033#discussion_r581354568 this won't actually work as intended and another solution is needed.

My fix in this PR aims to delay the actual deletion by some configurable value and the deletion logic will then either delete immediately or queue the deletion for after this period.

@terrytangyuan tagging you as you did the original PR and tagging @alexec as you were the reviewer, as you might have more insight here. I have written an initial test here but am happy to expand coverage if desired.

One area I am not sure about is what other docs need to be updated for an API change and if that is a manual process or if there is some automation around that, it was not clear to me as I could not find any documentation explaining how to make contributions that add to the spec and this is my first time so just let me know what else I need to do.

Checklist:

* [X] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -705,7 +705,13 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
 		}
 		if doPodGC {
 			for podName := range woc.completedPods {
-				woc.controller.queuePodForCleanup(woc.wf.Namespace, podName, deletePod)
+				if woc.execWf.Spec.PodGC.DeleteDelaySeconds != nil {
```
Comment (stefansedich) [inline workflow/controller/controller.go:708]: This logic is duplicated between here and the operator, I am usually in the "refactor after 3 times" camp but can be swayed to do something now.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -821,6 +821,8 @@ type PodGC struct {
 	Strategy PodGCStrategy `json:"strategy,omitempty" protobuf:"bytes,1,opt,name=strategy,casttype=PodGCStrategy"`
 	// LabelSelector is the label selector to check if the pods match the labels before being added to the pod GC queue.
 	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty" protobuf:"bytes,2,opt,name=labelSelector"`
+	// DeleteDelaySeconds is how long to delay pod deletion for.
+	DeleteDelaySeconds *int32 `json:"deleteDelaySeconds,omitempty" protobuf:"bytes,3,opt,name=deleteDelaySeconds"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:825]: I think this should be `uint32` or `metav1.Duration` not pointer type, zero would be delete ASAPA

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -705,7 +705,13 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
...
+					delayDuration := time.Duration(*woc.execWf.Spec.PodGC.DeleteDelaySeconds) * time.Second
+
+					woc.controller.queuePodForCleanupAfter(woc.wf.Namespace, podName, deletePod, delayDuration)
+				} else {
+					woc.controller.queuePodForCleanup(woc.wf.Namespace, podName, deletePod)
```
Comment (alexec) [inline workflow/controller/controller.go:713]: `queuePodForCleanup` === `queuePodForCleanupAfter(delayDuration=0)`

???

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 2 above)
Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:825]: That makes sense will update.

File: workflow/controller/controller.go (hunk 3 above)
Comment (stefansedich) [reply workflow/controller/controller.go:713]: Sounds fair, to be honest I wasn't entirely sure if the queue semantics so kept my changes as safe as possible but I will update.

**Comment by stefansedich** [comment]
@alexec have implemented your feedback, let me know what you think.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 4]
```diff
@@ -821,6 +821,8 @@ type PodGC struct {
 	Strategy PodGCStrategy `json:"strategy,omitempty" protobuf:"bytes,1,opt,name=strategy,casttype=PodGCStrategy"`
 	// LabelSelector is the label selector to check if the pods match the labels before being added to the pod GC queue.
 	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty" protobuf:"bytes,2,opt,name=labelSelector"`
+	// DeleteDelaySeconds is how long to delay pod deletion for.
+	DeleteDelaySeconds uint32 `json:"deleteDelaySeconds,omitempty" protobuf:"bytes,3,opt,name=deleteDelaySeconds"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:825]: I think this should be a general GC delay.

Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:825]: If we do that I assume we want to delay all GC actions including things like labelling pods? in my case I only care about delaying the deletion, I cannot think of a scenario where I would want to delay labelling the pod as complete.

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:825]: This only applies to `PodGCStrategy` - labeling is not really part of that. I think `DelaySeconds` would be good unless @jessesuen should we prefer `time.Duration` to `*Seconds`?

Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Thanks @alexec have changed to `DelaySeconds` will await further feedback from @jessesuen on if we prefer a Duration or not.

File: test/e2e/pod_cleanup_test.go [hunk 5]
```diff
@@ -477,6 +477,45 @@ spec:
 		})
 }
 
+func (s *PodCleanupSuite) TestPodDeletionDelay() {
```
Comment (terrytangyuan) [inline test/e2e/pod_cleanup_test.go:480]: Nice test!

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 4 above)
Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: @alexc @jessesuen any more feedback on this PR?

Comment (AlexC) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: @sgloutnikov  @alexec Heads up that I keep getting quite a lot of incorrect mentions from this repository from you guys .... it appears we have very similar usernames

Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: > @sgloutnikov  @alexec Heads up that I keep getting quite a lot of incorrect mentions from this repository from you guys .... it appears we have very similar usernames

Oops sorry my bad!

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 6]
```diff
@@ -821,6 +821,8 @@ type PodGC struct {
 	Strategy PodGCStrategy `json:"strategy,omitempty" protobuf:"bytes,1,opt,name=strategy,casttype=PodGCStrategy"`
 	// LabelSelector is the label selector to check if the pods match the labels before being added to the pod GC queue.
 	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty" protobuf:"bytes,2,opt,name=labelSelector"`
+	// DelaySeconds is how long to delay pod deletion for.
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:824]: should say "delay pod GC"

File: workflow/controller/operator.go [hunk 7]
```diff
@@ -589,13 +589,18 @@ func (woc *wfOperationCtx) persistUpdates(ctx context.Context) {
...
-					woc.controller.queuePodForCleanup(woc.wf.Namespace, podName, deletePod)
-				}
+				doPodGC = podPhase == apiv1.PodSucceeded
 			case wfv1.PodGCOnPodCompletion:
-				woc.controller.queuePodForCleanup(woc.wf.Namespace, podName, deletePod)
+				doPodGC = true
+			}
+
+			if doPodGC {
+				delayDuration := time.Duration(woc.execWf.Spec.PodGC.DelaySeconds) * time.Second
```
Comment (alexec) [inline workflow/controller/operator.go:602]: can you add a `func (p PodGC) GetDelay() time.Duration` to hide this in?

File: workflow/controller/operator.go [hunk 8]
```diff
@@ -589,13 +589,18 @@ func (woc *wfOperationCtx) persistUpdates(ctx context.Context) {
 	// errors if we label a pod that was deleted already.
 	for podName, podPhase := range woc.completedPods {
 		if woc.execWf.Spec.PodGC != nil {
+			var doPodGC bool
```
Comment (alexec) [inline workflow/controller/operator.go:592]: `do*` variables are kind of smelly, they tell your cause and effect have been separated in the code and risk bugs, instead just inline `queuePodForCleanupAfter` please

**Review by alexec — CHANGES_REQUESTED** [review]

File: workflow/controller/operator.go (hunk 8 above)
Comment (stefansedich) [reply workflow/controller/operator.go:592]: Okies will do, I had copied the pattern from elsewhere in the code but happy to oblige!

File: workflow/controller/operator.go (hunk 7 above)
Comment (stefansedich) [reply workflow/controller/operator.go:602]: Sure can makes sense.

**Comment by alexec** [comment]
bump! let me know if you want any help

**Comment by stefansedich** [comment]
> bump! let me know if you want any help

I had completely skipped on this :cry:  I am putting another PR together at the moment so will get this one cleaned up once I get that one complete.

**Comment by stefansedich** [comment]
> bump! let me know if you want any help

@alexec ok got this rebased and the changed implemented, should be ready to go.

**Comment by alexec** [comment]
Closed?

**Comment by stefansedich** [comment]
> Closed?

I had a rebase booboo, first time I made this much of a mess :laughing: will my PRs re-opened.

**Pull request closed without merging**